### PR TITLE
fix remove handler cause ByteToMessageDecoder out disorder

### LINKED ISSUE #9668: remove handler cause ByteToMessageDecoder out disorder
### remove handler cause ByteToMessageDecoder out disorder
	
	I input byte stream [1,2,3,4,5], when i read `4` then remove  ByteToMessageDecoder,
	 the out is stream [1,2,3,5,4]
	
	Here are the test cases，the cases will fail

```
 @Test
    public void testDisorder(){
        ByteToMessageDecoder decoder = new ByteToMessageDecoder() {
            int count = 0;

            //read 4 byte then remove this decoder
            @Override
            protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
                out.add(in.readBytes(1));
                if (++count >= 4) {
                    ctx.pipeline().remove(this);
                }
            }
        };

        EmbeddedChannel channel = new EmbeddedChannel(decoder);
        assertTrue(channel.writeInbound(Unpooled.wrappedBuffer(new byte[]{1, 2, 3, 4, 5})));
        assertEquals(1, ((ByteBuf) channel.readInbound()).readByte());
        assertEquals(2, ((ByteBuf) channel.readInbound()).readByte());
        assertEquals(3, ((ByteBuf) channel.readInbound()).readByte());
        assertEquals(4, ((ByteBuf) channel.readInbound()).readByte());
        assertEquals(5, ((ByteBuf) channel.readInbound()).readByte());
        assertFalse(channel.finish());
    }


```
### window10

### PR DESCRIPTION
**Pull request by switchYello**
Motivation:

Data flowing in from the decoder flows out in sequence，Whether decoder removed or not.

Modification:

fire data in out and clear out when hander removed
before call method handlerRemoved(ctx)

Result:

Fixes #9668 .

### CODE REVIEW

File: codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java [hunk 1]
```diff
@@ -507,6 +507,8 @@ final void decodeRemovalReentryProtection(ChannelHandlerContext ctx, ByteBuf in,
             boolean removePending = decodeState == STATE_HANDLER_REMOVED_PENDING;
             decodeState = STATE_INIT;
             if (removePending) {
+                fireChannelRead(ctx,out,out.size());
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java]: nit: add whitespace after each `,`

File: codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java [hunk 2]
```diff
@@ -399,4 +399,29 @@ public void read(ChannelHandlerContext ctx) throws Exception {
...
+            }
+        };
+
+        EmbeddedChannel channel = new EmbeddedChannel(decoder);
+        assertTrue(channel.writeInbound(Unpooled.wrappedBuffer(new byte[]{1, 2, 3, 4, 5})));
+        assertEquals(1, ((ByteBuf) channel.readInbound()).readByte());
+        assertEquals(2, ((ByteBuf) channel.readInbound()).readByte());
+        assertEquals(3, ((ByteBuf) channel.readInbound()).readByte());
+        assertEquals(4, ((ByteBuf) channel.readInbound()).readByte());
+        assertEquals(5, ((ByteBuf) channel.readInbound()).readByte());
```
Comment (normanmaurer) [inline codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java]: You will need to call `release()` on each of the `ByteBuf` instances... Alternative you could change your decoder to just do:

```
out.add(in.readByte());
```

And then assert the returned `byte`.

**Comment by normanmaurer** [comment]
@switchYello ping me once you addressed the comments

**Comment by switchYello** [comment]
@normanmaurer  yes

**Comment by normanmaurer** [comment]
@johnou PTAL

**Comment by johnou** [comment]
@normanmaurer what about https://github.com/netty/netty/issues/4635 ? or do you see this as another edge case?

**Comment by normanmaurer** [comment]
@switchYello please fix the check style errors:

```
15:17:12 [INFO] Starting audit...
15:17:12 [ERROR] /code/codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java:404:31: '{' is not preceded with whitespace. [WhitespaceAround]
15:17:12 [ERROR] /code/codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java:406:17: Variable 'count' explicitly initialized to '0' (default value for its type). [ExplicitInitialization]
15:17:12 [ERROR] /code/codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java:419:27: 'typecast' is not followed by whitespace. [WhitespaceAfter]
15:17:12 [ERROR] /code/codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java:420:27: 'typecast' is not followed by whitespace. [WhitespaceAfter]
15:17:12 [ERROR] /code/codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java:421:27: 'typecast' is not followed by whitespace. [WhitespaceAfter]
15:17:12 [ERROR] /code/codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java:422:27: 'typecast' is not followed by whitespace. [WhitespaceAfter]
15:17:12 [ERROR] /code/codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java:424:27: 'typecast' is not followed by whitespace. [WhitespaceAfter]
15:17:12 Audit done.
```

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by switchYello** [comment]
@normanmaurer 

the output must be byte,i conversion to byte fail，I don't understand

```
12:07:22 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:testCompile (default-testCompile) on project netty-codec: Compilation failure: Compilation failure: 
12:07:22 [ERROR] /code/codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java:[419,57] error: incompatible types: Object cannot be converted to byte
12:07:22 [ERROR] /code/codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java:[420,57] error: incompatible types: Object cannot be converted to byte
12:07:22 [ERROR] /code/codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java:[421,57] error: incompatible types: Object cannot be converted to byte
12:07:22 [ERROR] /code/codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java:[422,57] error: incompatible types: Object cannot be converted to byte

```

if i not conversion ,my idea will prompt me

ideal

and now I don't know how to do it. remove type conversion ?

**Comment by switchYello** [comment]
oh i konw ,because my languagel level is 1.8 ... ,netty need 1.6

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@johnou sorry I don't get what you are asking for... Can you elaborate ?

File: codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java [hunk 3]
```diff
@@ -399,4 +399,30 @@ public void read(ChannelHandlerContext ctx) throws Exception {
...
+            }
+        };
+        EmbeddedChannel channel = new EmbeddedChannel(decoder);
+        assertTrue(channel.writeInbound(Unpooled.wrappedBuffer(new byte[]{1, 2, 3, 4, 5})));
+        assertEquals((byte) 1,  channel.readInbound());
+        assertEquals((byte) 2,  channel.readInbound());
+        assertEquals((byte) 3,  channel.readInbound());
+        assertEquals((byte) 4,  channel.readInbound());
+        ByteBuf buffer5 = channel.readInbound();
+        assertEquals((byte) 5, buffer5.readByte());
```
Comment (normanmaurer) [inline codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java]: nit: can we also add `assertFalse(buffer5.isReadable());` here to ensure we not produce more then expected ?

Comment (normanmaurer) [reply codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java]: @switchYello let me know once you did address this and I will merge.

**Comment by switchYello** [comment]
@normanmaurer  yes

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@switchYello thanks a lot!
